core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/protocol/restxml/XmlProtocolMarshaller.java

```diff
@@ -0,0 +1,179 @@
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.DefaultRequest;
+import software.amazon.awssdk.core.Request;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.protocol.OperationInfo;
+import software.amazon.awssdk.core.protocol.ProtocolMarshaller;
+import software.amazon.awssdk.core.protocol.SdkField;
+import software.amazon.awssdk.core.protocol.SdkPojo;
+import software.amazon.awssdk.core.protocol.traits.DefaultValueTrait;
+import software.amazon.awssdk.core.protocol.traits.PayloadTrait;
+import software.amazon.awssdk.core.util.UriResourcePathUtils;
+import software.amazon.awssdk.utils.Md5Utils;
+import software.amazon.awssdk.utils.StringInputStream;
+
+@SdkInternalApi
+public class XmlProtocolMarshaller<OrigRequestT extends SdkRequest> implements ProtocolMarshaller<Request<OrigRequestT>> {
+
+    private static final MarshallerRegistry MARSHALLER_REGISTRY = createMarshallerRegistry();
+
+    private final Request<OrigRequestT> request;
+    private final XmlMarshallerContext marshallerContext;
+
+    public XmlProtocolMarshaller(XmlGenerator xmlGenerator,
+                                 OrigRequestT originalRequest,
+                                 OperationInfo operationInfo) {
+        this.request = fillBasicRequestParams(operationInfo, originalRequest);
+        this.marshallerContext = XmlMarshallerContext.builder()
+                                                     .xmlGenerator(xmlGenerator)
+                                                     .marshallerRegistry(MARSHALLER_REGISTRY)
+                                                     .protocolMarshaller(this)
+                                                     .request(request)
+                                                     .build();
+    }
+
+    private Request<OrigRequestT> fillBasicRequestParams(OperationInfo operationInfo, OrigRequestT originalRequest) {
+        Request<OrigRequestT> request = new DefaultRequest<>(originalRequest, operationInfo.serviceName());
+        request.setHttpMethod(operationInfo.httpMethodName());
+        request.setResourcePath(UriResourcePathUtils.addStaticQueryParametersToRequest(request, operationInfo.requestUri()));
+        return request;
+    }
+
+    @Override
+    public Request<OrigRequestT> marshall(SdkPojo pojo) {
+        doMarshall(pojo);
+        return finishMarshalling();
+    }
+
+    public void doMarshall(SdkPojo pojo) {
+        for (SdkField<?> field : pojo.sdkFields()) {
+            Object val = resolveValue(field.get(pojo), field);
+
+            if (isBinary(field, val)) {
+                request.setContent(((SdkBytes) val).asInputStream());
+                setContentTypeHeaderIfNeeded("binary/octet-stream");
+
+            } else if (isExplicitPayloadMember(field) && val instanceof String) {
```

Comment:
This branch seems very S3 specific. Is there anything we can do to keep it separate?